libraries/stdlib/src/kotlin/io/Console.kt

```diff
@@ -132,50 +136,89 @@ public inline fun println() {
     System.out.println()
 }
 
-// Since System.in can change its value on the course of program running,
-// we should always delegate to current value and cannot just pass it to InputStreamReader constructor.
-// We could use "by" implementation, but we can only use "by" with interfaces and InputStream is abstract class.
-private val stdin: BufferedReader by lazy { BufferedReader(InputStreamReader(object : InputStream() {
-    public override fun read(): Int {
-        return System.`in`.read()
-    }
+private const val BUFFER_SIZE: Int = 32
 
-    public override fun reset() {
-        System.`in`.reset()
-    }
+// Since System.in can change its value on the course of program running, we should always delegate to current value.
+private val stdin: InputStream
+    get() = System.`in`
 
-    public override fun read(b: ByteArray): Int {
-        return System.`in`.read(b)
-    }
+private var cachedDecoder: CharsetDecoder? = null
 
-    public override fun close() {
-        System.`in`.close()
+private fun decoderFor(charset: Charset): CharsetDecoder {
+    return cachedDecoder?.takeIf { cached ->
+        cached.charset() == charset
+    } ?: charset.newDecoder().also { newDecoder ->
+        cachedDecoder = newDecoder
     }
+}
 
-    public override fun mark(readlimit: Int) {
-        System.`in`.mark(readlimit)
+/**
+ * Reads a line of input from the standard input stream.
+ *
+ * @return the line read or `null` if the input stream is redirected to a file and the end of file has been reached.
+ */
+fun readLine(lineSeparator: String = System.lineSeparator(), charset: Charset = Charset.defaultCharset()): String? =
```

Comment:
It should work with both types of line separators automatically, and there's no requirement to work with custom ones, so the `lineSeparator` parameter is not needed.